inspire_dojson/api.py

```diff
@@ -0,0 +1,151 @@
+from inspire_utils.record import get_value
+
+from .conferences import conferences
+from .data import data
+from .experiments import experiments
+from .hep import hep, hep2marc
+from .hepnames import hepnames, hepnames2marc
+from .institutions import institutions
+from .jobs import jobs
+from .journals import journals
+
+RECORD = E.record
+CONTROLFIELD = E.controlfield
+DATAFIELD = E.datafield
+SUBFIELD = E.subfield
+
+
+def marcxml2record(marcxml):
+    """Convert a MARCXML string to a JSON record.
+
+    Tries to guess which set of rules to use by inspecting the contents
+    of the ``980__a`` MARC field, but falls back to HEP in case nothing
+    matches, because records belonging to special collections logically
+    belong to the Literature collection but don't have ``980__a:HEP``.
+
+    Args:
+        marcxml(str): a string containing MARCXML.
+
+    Returns:
+        dict: a JSON record converted from the string.
+
+    """
+    marcjson = create_record(marcxml, keep_singletons=False)
+    collections = _get_collections(marcjson)
+
+    if 'conferences' in collections:
+        return conferences.do(marcjson)
+    elif 'data' in collections:
+        return data.do(marcjson)
+    elif 'experiment' in collections:
+        return experiments.do(marcjson)
+    elif 'hepnames' in collections:
+        return hepnames.do(marcjson)
+    elif 'institutions' in collections:
+        return institutions.do(marcjson)
+    elif 'job' in collections or 'jobhidden' in collections:
+        return jobs.do(marcjson)
+    elif 'journals' in collections or 'journalsnew' in collections:
+        return journals.do(marcjson)
+    return hep.do(marcjson)
+
+
+def record2marcxml(record):
+    """Convert a JSON record to a MARCXML string.
+
+    Deduces which set of rules to use by parsing the ``$schema`` key, as
+    it unequivocally determines which kind of record we have.
+
+    Args:
+        record(dict): a JSON record.
+
+    Returns:
+        str: a MARCXML string converted from the record.
+
+    """
+    schema_name = _get_schema_name(record)
+
+    if schema_name == 'hep':
+        marcjson = hep2marc.do(record)
+    elif schema_name == 'hepnames':
+        marcjson = hepnames2marc.do(record)
+    else:
+        raise NotImplementedError(u'JSON -> MARC rules not implemented for {}'.format(schema_name))
+
+    record = RECORD()
+
+    for key, values in sorted(iteritems(marcjson)):
+        tag, ind1, ind2 = _parse_key(key)
+        if _is_controlfield(tag, ind1, ind2):
+            value = force_single_element(values)
+            if not isinstance(value, text_type):
+                value = text_type(value)
+            record.append(CONTROLFIELD(value, {'tag': tag}))
+        else:
+            datafield = DATAFIELD({'tag': tag, 'ind1': ind1, 'ind2': ind2})
+            for value in force_list(values):
+                for code, el in sorted(iteritems(value)):
+                    if not isinstance(el, text_type):
```

Comment:
is this handling repeated subfields, like multiple affiliations for one author?

Reply:
Right, I didn't, so I added a test for this: https://github.com/inspirehep/inspire-dojson/pull/160/files#diff-625efbbf35a2290861f596fee1b19838R180.